File: Commons/Action/src/main/java/com/greentree/commons/action/PoolAction.java
package com.greentree.commons.action;

import java.util.Collection;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

import com.greentree.commons.action.observer.object.IObjectAction;


public class PoolAction<T, A extends IObjectAction<T>> implements IObjectAction<T> {
	
	private static final long serialVersionUID = 1L;
	
	private final Collection<T> pool = new CopyOnWriteArrayList<>();
	private final A action;
	
	public PoolAction(A action) {
		this.action = action;
	}
	
	@Override
	public ListenerCloser addListener(Consumer<? super T> l) {
		final var res = action.addListener(l);
		for(var v : pool)
			l.accept(v);
		return res;
	}
	
	@Override
	public void clear() {
		clearPool();
		action.clear();
	}
	
	public void clearPool() {
		pool.clear();
	}
	
	@Override
	public void event(T t) {
		if(pool.contains(t))
			return;
		pool.add(t);
		action.event(t);
	}
	
	public A getAction() {
		return action;
	}
	
	@Override
	public int listenerSize() {
		return action.listenerSize();
	}
	
	@Override
	public String toString() {
		var builder = new StringBuilder();
		builder.append("PoolAction [");
		builder.append(action);
		builder.append("]");
		return builder.toString();
	}
	
}

File: Commons/Action/src/main/java/com/greentree/commons/action/PoolAction.kt
package com.greentree.commons.action

import com.greentree.commons.action.observer.`object`.IObjectAction
import java.util.concurrent.CopyOnWriteArrayList

open class PoolAction<T, A : IObjectAction<T>>(val action: A) : IObjectAction<T> {

	private val pool: MutableCollection<T> = CopyOnWriteArrayList<T>()

	override fun addListener(listener: Function1<T, Unit>): ListenerCloser {
		val res: ListenerCloser = action.addListener(listener)
		for(v in pool)
			listener.invoke(v)
		return res
	}

	override fun clear() {
		clearPool()
		action.clear()
	}

	fun clearPool() {
		pool.clear()
	}

	override fun event(t: T) {
		if(pool.contains(t)) return
		pool.add(t)
		action.event(t)
	}

	override val listenerCount: Int
		get() = action.listenerCount

	override fun toString(): String {
		val builder = "PoolAction [" +
				action +
				"]"
		return builder
	}

	override fun close() {
		pool.clear()
		action.close()
	}

	companion object {

		private const val serialVersionUID = 1L
	}
}
